### PR TITLE
feat: theorems about Ctxt and MVar

### PR DESCRIPTION
**Pull request by tobiasgrosser**
These theorems will allow us to reason about the context and MVar with less unfolding about low-level theorems. While uses of these theorems are still worked on, the theorems seem to be generally useful so lets upstream them already to keep later patches small.

Co-Authored-By : Siddharth Bhat<siddu.druid@gmail.com>

These are factored out of https://github.com/opencompl/ssa/pull/225.

### CODE REVIEW

**Comment by bollu** [comment]
@tobiasgrosser LGTM, and to stick to @goens ' principles, I add myself to the  `Co-Authored-By` set.

File: SSA/Core/Util/ConcreteOrMVar.lean [hunk 1]
```diff
@@ -26,6 +26,12 @@ instance : OfNat (ConcreteOrMVar Nat φ) n := ⟨.concrete n⟩
 
 namespace ConcreteOrMVar
 
+/- Terse notation for concrete and mvar. -/
+namespace Notation
+scoped notation "i" n => ConcreteOrMVar.concrete n
+scoped notation "i?" f => ConcreteOrMVar.mvar f
```
Comment (bollu) [inline SSA/Core/Util/ConcreteOrMVar.lean:32]: I'm not totally sold on the value of these notations, though I write them myself. I wanted to mimic LLVM syntax, so it looks like `i 1` and `i? 0` to stand for concrete numbers and mvars, but I'm not a huge fan. @alexkeizer 's second opinion is much appreciated.

**Comment by bollu** [comment]
LGTM from my end.

File: SSA/Core/Util/ConcreteOrMVar.lean (hunk 1 above)
Comment (tobiasgrosser) [reply SSA/Core/Util/ConcreteOrMVar.lean]: I dropped the notation for now, given that you are uncertain here.

**Comment by tobiasgrosser** [comment]
> @tobiasgrosser LGTM, and to stick to @goens ' principles, I add myself to the `Co-Authored-By` set.

I feel this needs to actually be part of the commit. I force pushed a commit with the correct tag and will push now. Thank you for the reminder.

File: SSA/Core/Util/ConcreteOrMVar.lean [hunk 2]
```diff
@@ -48,5 +48,22 @@ def instantiate (as : Vector α φ) : ConcreteOrMVar α φ → α
...
+  simp [instantiate]
+  simp [Vector.get]
+
+@[simp]
+lemma instantiate_mvar_zero' :
+    (mvar (φ := 1) ⟨0, by simp⟩).instantiate (Subtype.mk [w] (by simp)) = w := rfl
+
+@[simp]
+lemma instantiate_mvar_zero'' :
+    (mvar (φ := 1) 0).instantiate (Subtype.mk [w] h1) = w := rfl
```
Comment (alexkeizer) [inline SSA/Core/Util/ConcreteOrMVar.lean:67]: The difference between these three lemmas is not very clear to me, and obfuscated by different spellings of identical concepts (i.e., `ConcreterOrMVar.instantiate` vs dot-notation `(...).instantiate` and `Subtype.mk` vs magic braces `⟨_, _⟩`). I'd like to see this be more uniform, and I'd like to see a docstring to explain the differences

File: SSA/Core/ErasedContext.lean [hunk 3]
```diff
@@ -56,6 +56,16 @@ def get? : Ctxt Ty → Nat → Option Ty :=
 def map (f : Ty₁ → Ty₂) : Ctxt Ty₁ → Ctxt Ty₂ :=
   List.map f
 
+@[simp]
+lemma map_nil (f : Ty → Ty') :
+  map f ∅ = ∅ := rfl
+
+@[simp]
+lemma map_cons (Γ : Ctxt Ty) (t : Ty) (f : Ty → Ty') :
+  map f (Γ.cons t) = (Γ.map f).cons (f t) := rfl
```
Comment (alexkeizer) [inline SSA/Core/ErasedContext.lean:65]: There's some inconsistent use of dot-notation here (`map f _` vs `_.map f`).

Also, I think `cons` here is referring to `List.cons`, which we've alliased as `Ctxt.snoc`.
In the spirit of remaining higher-level, I think we should not be adding lemmas for `List.snoc`, to nudge us towards not unfolding `Ctxt.snoc` in the first place

**Review by alexkeizer — COMMENTED** [review]
I missed this PR before it was merged, but still wanted to share some thoughts

**Pull request merged**